web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts

```diff
@@ -0,0 +1,240 @@
+      // if we return nothing here, the file transfer will not be added to the
+      // file transfer list. If we add it to the list, the file will continue to
+      // start the download and return another here. This prevents a second network
+      // request that we know will fail.
+      return;
+    }
+    return getHttpFileTransferHandlers().download(url, abortController);
+  }
+
+  async function handleDownload(
+    location: string,
+    abortController: AbortController
+  ) {
+    if (session.moderated) {
+      tty.sendFileDownloadRequest(location);
+      return;
+    }
+
+    return download(location, abortController);
+  }
+
+  async function upload(
+    location: string,
+    file: File,
+    abortController: AbortController,
+    moderatedSessionParams?: ModeratedSessionParams
+  ) {
+    const url = await getScpUrl({
+      location,
+      clusterId: currentDoc.clusterId,
+      serverId: currentDoc.serverId,
+      login: currentDoc.login,
+      filename: file.name,
+      moderatedSessonId: moderatedSessionParams?.moderatedSessionId,
+      fileTransferRequestId: moderatedSessionParams?.fileRequestId,
+    });
+    if (!url) {
+      // if we return nothing here, the file transfer will not be added to the
+      // file transfer list. If we add it to the list, the file will continue to
+      // start the download and return another here. This prevents a second network
+      // request that we know will fail.
+      return;
+    }
+    return getHttpFileTransferHandlers().upload(url, file, abortController);
+  }
+
+  async function handleUpload(
+    location: string,
+    file: File,
+    abortController: AbortController
+  ) {
+    if (session.moderated) {
+      tty.sendFileUploadRequest(location, file);
+      return;
+    }
+
+    return upload(location, file, abortController);
+  }
+
+  function handleFileTransferDenied(request: FileTransferRequest) {
+    removeFileTransferRequest(request.requestID);
+  }
+
+  function handleFileTransferApproval(
+    request: FileTransferRequest,
+    file?: File
+  ) {
+    removeFileTransferRequest(request.requestID);
+    if (request.requester !== user.username) {
+      return;
+    }
+
+    if (request.download) {
+      return filesStore.start({
+        name: request.location,
+        runFileTransfer: abortController =>
+          download(request.location, abortController, {
+            fileRequestId: request.requestID,
+            moderatedSessionId: request.sid,
+          }),
+      });
+    }
+
+    // if it gets here, it's an upload
+    if (!file) {
+      throw new Error('Approved file not found for upload.');
+    }
+    return filesStore.start({
+      name: request.filename,
+      runFileTransfer: abortController =>
+        upload(request.location, file, abortController, {
+          fileRequestId: request.requestID,
+          moderatedSessionId: request.sid,
+        }),
+    });
+  }
+
+  function updateFileTransferRequests(data: FileTransferRequest) {
```

Comment:
I won't move them into separate hooks but i moved things around a bit. I added your comment blocks and also moved the tty event listeners up to right under the `useEffect` hook where they are bound. I also added a comment to each one to describe what it's handling. It may include a bit more backend info than needed but hey, not a huge problem to have a bit more info.

> Though I have to say, it'd make it much easier to under what all those functions are doing! But I know that this is like a round three of code review so I understand if making further changes is just tiresome.

Code review is never tiresome for this reviewee!